### ISSUE TITLE
Files uploaded to external storage not shown immediately in Webinterface in shared folders

### ISSUE THREAD

**Issue by rgl1234** [nextcloud/server#23988]
Hello
There seems to be a problem in NC20.01 and external storages. When I place new files on external storage, they are not shown in WebInterface as normal. I get this Problem after upgrading from 19.04 to NC20.01.

**Comment by mmaridev** [comment]
Can confirm, this issue was firstly reported by me and my team for Nextcloud 18. I just tried again and can confirm is still reproducible.

Steps to reproduce:
1. Add an external storage in Nextcloud
2. Create a directory from Nextcloud and check via shell/SMB/NFS/whatever your storage: the directory has been created
3. Via another protocol (NOT Nextcloud) create a file in the directory: the file won't show in the GUI

Current workaround:
Periodically run
```
DELETE FROM ncfilecache WHERE storage NOT IN (SELECT f.numeric_id FROM ncstorages f);
VACUUM FULL ncstorages;
VACUUM FULL ncfilecache;
DELETE FROM ncstorages;
```
in your db

**Comment by rgl1234** [comment]
peoblem still present in NC 20.02

**Comment by ChewingBever** [comment]
Can confirm this problem is still present in NC 20.04

**Comment by rgl1234** [comment]
unfortunately still present in NC 20.05
> does someone have an idea what changed here from V19 to V20? It workes correctly in V19.....

**Comment by rgl1234** [comment]
I could now find maybe the source of the problem. In external storages there is the checkbox
"check for changes" > "once every direct access"
In NC20, this functionality is only working, when the NC-user has direct access to the NC-Folder. But when the NC-user has access to the folder by "shared-folder", the "check for changes"-function is not executed and the user cannot see the correct content of the folder....

Can someone confirm this bug in NC20?

**Comment by quizilkend** [comment]
> I could now find maybe the source of the problem. In external storages there is the checkbox
> "check for changes" > "once every direct access"
> In NC20, this functionality is only working, when the NC-user has direct access to the NC-Folder. But when the NC-user has access to the folder by "shared-folder", the "check for changes"-function is not executed and the user cannot see the correct content of the folder....
> 
> Can someone confirm this bug in NC20?

Thanks for writing this up. I experience exactly the same on NC20.0.5

**Comment by ikarlo** [comment]
> Thanks for writing this up. I experience exactly the same on NC20.0.5

+1

**Comment by rgl1234** [comment]
Does someone know how to proceed with this bug to have it fixed in next release?

**Comment by adrfantini** [comment]
Just here to confirm this behaviour in the latest version from the official docker, using local storage. Can reproduce 100% of the time.

**Comment by Flunkyball** [comment]
+1
NC 20.0.2

**Comment by addagrm** [comment]
> 
> 
> I could now find maybe the source of the problem. In external storages there is the checkbox
> "check for changes" > "once every direct access"
> In NC20, this functionality is only working, when the NC-user has direct access to the NC-Folder. But when the NC-user has access to the folder by "shared-folder", the "check for changes"-function is not executed and the user cannot see the correct content of the folder....
> 
> Can someone confirm this bug in NC20?

I see this issue in Nextcloud 21.0.0 as well as in 20.0.8. Refresh fails as soon as an external storage is shared with another user.

**Comment by rgl1234** [comment]
thanks @addagrm for your feedback. So I hope someone from nextcloud will have a look into this problem soon...

**Comment by Flunkyball** [comment]
@addagrm 
Is the refresh working if the external storage is not shared ? My users have their home directory included as external storage via SFTP (tried SMB as well) and experience the same issue - no refresh what so ever.

**Comment by addagrm** [comment]
@Flunkyball 
Refresh is working for a local directory when the original user accesses the directory. I don't use SFTP or SMB as external storage.

**Comment by rgl1234** [comment]
@Flunkyball 
in my case I use SMB as external storage. Refreshing is only working, when original user is accessing the folder ("check for changes" on folder set to "once every direct access"). If the shared user is accessing the same folder, refreshing is not working :-(

**Comment by Flunkyball** [comment]
@rgl1234 
Regarding SMB I just came across an occ command to enable listening for updates on SMB shares:
[https://docs.nextcloud.com/server/stable/admin_manual/configuration_files/external_storage/smb.html#smb-update-notifications](url)
Have you tried that already ?

In my case users access their own home directory via SFTP or SMB. Through both ways changes do not show up (except in some random occasions; once every 100 times) in Nextcloud web ui or app. 
Including the home directories via local mount is not an option cause according to the documentation www-data needs to be the owner of the directory and the chmod level is 750 which does not work for private home directories.

**Comment by rgl1234** [comment]
@Flunkyball 
Unfortunately in my case smb-update-notifications do not make any difference. Its also noted, that it only works with windows smb-server. I use smb on linux (ubuntu). The problem starts after upgrading to V20. So until know I am not able to upgrade my V19-version because of that bug....its realy a bug of a key-feature, that will hopefully be solved soon....

**Comment by rgl1234** [comment]
Bug still present in V21.0.1 (tested with official nextcloud:fpm docker-image)

**Comment by addagrm** [comment]
In the meanwhile I implemented an additional script to rescan based on watch. It has some downsides but is a first workaround. I think such a script wont work for SMB setups.

**Comment by Flunkyball** [comment]
@addagrm 
Sounds interesting anyway. Am I assuming that you monitoring file access on a file system level and trigger a rescan via ocs if detected ? Do you want to share some more information ?

**Comment by q-wertz** [comment]
> I could now find maybe the source of the problem. In external storages there is the checkbox
> "check for changes" > "once every direct access"
> In NC20, this functionality is only working, when the NC-user has direct access to the NC-Folder. But when the NC-user has access to the folder by "shared-folder", the "check for changes"-function is not executed and the user cannot see the correct content of the folder....
> 
> Can someone confirm this bug in NC20?

We experienced the same issue.

In our case the external storage is "mounted" in a subfolder of a NC Group folder

Some more things we found out when investigating the issue:
- The user has to be have access in the Group folder AND the external storage folder to trigger a refresh
- Nice thing: when I (having permission to enter the folder directly) enter the folder it is refreshed and he also can see the files
- Even after everything works because he is added to the GroupFolder and the External one, when he accesses the folder through the shared folder it is still not updated

**Comment by rgl1234** [comment]
hello @q-wertz 
Thanks for your input. But how did you mount an external storage into a group-folder?

**Comment by q-wertz** [comment]
I created a group folder e.g. `GroupFolder` and shared it e.g. with `Group1` and `Group2` then in the external storage settings I added two sftp Folders which are named in the interface `GroupFolder/ExStorage01` and `GroupFolder/ExStorage02`

 `ExStorage01` is accessible for `Group1`
 `ExStorage02` is accessible for `Group2`

Everything works perfectly except,that when e.g. someone from `Grou02` shares one of the subfolders of `ExStorage02` with someone who is not in `Group02`. When this person accesses the folder, no refresh of the files list in this folder is triggered.

I think that's a general issue with the shared folder mechanism. It does not trigger this filescan that is triggered by everyone that has the access (granted by the External Storage App) to the folder.
As I described, when someone from `Group2` enters the folder, the refresh is triggered and now also the person with the usual Nextcloud share (but not part of `Group2`) can see the changes that happened in the meantime.

**Comment by rgl1234** [comment]
thanks @q-wertz for the clarification. It is exactly the same behaviour as we face..... This bug was introduced with V20 and is still present in V21. In V19 everithing is working correctly
Hope that someone from nextcloud can invest into this one day......

**Comment by twocore** [comment]
We are using Nextcloud 20.0.x for our customers and can confirm this behavior. Is there a solution or a workaround because our customers are dissatisfied?

**Comment by mmaridev** [comment]
@twocore if they don't care about shares, see my workaround in https://github.com/nextcloud/server/issues/23988#issuecomment-724132990

**Comment by adrfantini** [comment]
I'm using 21.0.1 and for some reason this bug does not affect me anymore. At least, the last 3 times I added files to my external storage (from NFS, not from NC), they were properly recognized in NC immediately.

**Comment by rgl1234** [comment]
I am also using 21.0.1 but unfortunately bug still present (NC docker image)

**Comment by q-wertz** [comment]
> I'm using 21.0.1 and for some reason this bug does not affect me anymore. At least, the last 3 times I added files to my external storage (from NFS, not from NC), they were properly recognized in NC immediately.

@adrfantini Do you have access to the folder directly (via External Storage settings) or been granted access by a share?

**Comment by adrfantini** [comment]
> > I'm using 21.0.1 and for some reason this bug does not affect me anymore. At least, the last 3 times I added files to my external storage (from NFS, not from NC), they were properly recognized in NC immediately.
> 
> @adrfantini Do you have access to the folder directly (via External Storage settings) or been granted access by a share?

The former, I have direct access to this External Storage as this is my private server.

**Comment by q-wertz** [comment]
> The former, I have direct access to this External Storage as this is my private server.

Ah ok. Never had an issue there (or never noticed), only issue we have is with shared folders from the External Folder.

**Comment by twocore** [comment]
@mmaridev Unfortunately, the workaround doesn't work for me. After executing the SQL commands, the users with shared folders can no longer see any data. I must login with the user, which has direct access to the external share (SMB).

Sorry for my english.

**Comment by rgl1234** [comment]
Hello, dows maybe someone has the possibility to test this Issue in NC22 Beta ? I hope it can be solved for NC22 final....
#24287

**Comment by adrfantini** [comment]
> I'm using 21.0.1 and for some reason this bug does not affect me anymore. At least, the last 3 times I added files to my external storage (from NFS, not from NC), they were properly recognized in NC immediately.

Unfortunately the issue came back. 3 hours since I uploaded a new photo folder and it is not showing up. I have not updated my NC installation in the meantime, so it working (or not working) was possibly just a fluke :/

**Comment by szaimen** [comment]
> In NC20, this functionality is only working, when the NC-user has direct access to the NC-Folder. But when the NC-user has access to the folder by "shared-folder", the "check for changes"-function is not executed and the user cannot see the correct content of the folder....

I suppose this is still happening on NC21.0.4?

**Comment by rgl1234** [comment]
thanks @szaimen 
yes, still present in NC21.0.4. I just tried it in official-docker-image of 21.0.4.

**Comment by 4pack** [comment]
I'll see if I can spin up a duplicate(ish) install that I can test that on. I also answered this part myself. This might help someone resolve this temporarily.

> Since most of the users have the same external storage, is it possible to just run the files:scan command for a single user and have the updates propagate to the rest of them, or is external storage coded in Nextcloud where each user has their own set of file IDs for external storage?

No it does not, you need to sync every user that has access to external storage. Definitely not ideal if you have a lot of users.

**Comment by ib-mlatin** [comment]
I'll see if I can spin up a duplicate(ish) install that I can test that on. I also answered this part myself. This might help someone resolve this temporarily.

> Since most of the users have the same external storage, is it possible to just run the files:scan command for a single user and have the updates propagate to the rest of them, or is external storage coded in Nextcloud where each user has their own set of file IDs for external storage?

No it does not, you need to sync every user that has access to external storage. Definitely not ideal if you have a lot of users.

**Comment by rgl1234** [comment]
thanks @szaimen 
I just tested the fix in NC21 and can confirm, that the problem seems to be solved.....

**Comment by ciberboy** [comment]
> 2\. filesystem_check_changes: 1

Hi

Where did you set this option? filesystem_check_changes: 1

**Comment by mmaridev** [comment]
@ciberboy 
> Where did you set this option? filesystem_check_changes: 1

via `occ` command, see [this link](https://gitlab.fuss.bz.it/fuss-team/fuss-nc/-/blob/master/nextcloud-fuss/configure-fuss.sh#L126).

**Comment by rgl1234** [comment]
hello @szaimen 
Unfortunately, this bug is back again after having updated my DB from V25 to V26.0.1
So I think another fix is needed for V26.0.1

**Comment by rgl1234** [comment]
hello @szaimen 
Bug still available in V27.0.0

**Comment by szaimen** [comment]
Hi, please open a new bug report then with up-to-date information.
